#!/applications/R/R-3.4.0/bin/Rscript
### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0 (invoke this version by specifying path at command line:
# /applications/R/R-3.4.0/bin/R , or if running as a script:
# /applications/R/R-3.4.0/bin/Rscript)

# Usage: tximport_extract_all_geneIDs.R ../samples_no_Col_Rep2.txt DESeq2_analysis_02

args <- commandArgs(trailingOnly = TRUE)
sampleFile <- args[1]
#sampleFile <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/samples_no_Col_Rep2.txt"
analysisDir <- args[2]
#analysisDir <- "DESeq2_analysis_02"

outDir <- paste0(dirname(sampleFile), "/", analysisDir, "/")

library(tximport)
print(packageVersion("tximport"))
#[1] ‘1.4.0’

# Read in table of sample IDs that will be used to specify paths to count files
samples <- read.table(sampleFile, header = T)
print(samples)

# Specify paths to count files
files <- file.path(dirname(sampleFile), samples$genotype, samples$directory, samples$sample, "quant.sf")
# Set 1:#_samples
names(files) <- paste0("sample", 1:8)
all(file.exists(files))
#[1] TRUE

# Create a dataframe of transcript IDs and corresponding gene IDs
transID <- read.table(files[1], colClasses = c(NA, rep("NULL", 4)), header = T)
tx2gene <- data.frame(cbind(as.vector(transID[,1]), substr(transID[,1], 1, 9)))
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Import transcript-level counts, summarised at gene level
# (reads that map to transcript IDs with a common parent gene ID are pooled)
library(readr)
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
print(names(txi))
#[1] "abundance"           "counts"              "length"             
#[4] "countsFromAbundance"
TAIR10_geneIDs <- rownames(txi$abundance)
TAIR10_geneIDsChr <- TAIR10_geneIDs[!grepl("ATM", TAIR10_geneIDs) &
                                    !grepl("ATC", TAIR10_geneIDs)]
save(TAIR10_geneIDsChr, file = paste0(outDir, "TAIR10_geneIDsChr.RData"))


