#!/applications/R/R-3.4.0/bin
### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0 (invoke this version by specifying path at command line: /applications/R/R-3.4.0/bin/R , or if running as a script: /applications/R/R-3.4.0/bin/Rscript)
# DESeq2 version 1.16.1
# Note that this R version or later is required for DESeq2 version 1.16 or later,
# in which "the log2 fold change shrinkage is no longer default for the DESeq and nbinomWaldTest functions".
# DESeq2 version 1.16 introduces "a separate function lfcShrink, which performs log2 fold change shrinkage
# for visualization and ranking of genes." (see https://support.bioconductor.org/p/95695/ and
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#changes)

## Import transcript abundance datasets with tximport package
# Code below based on https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

library(tximport)
print(packageVersion("tximport"))
#[1] ‘1.4.0’

inDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/"
outDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/"
plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/"

# Read in table of sample IDs that will be used to specify paths to count files
samples <- read.table(file.path(inDir, "samples.txt"), header = T)
print(samples)
#  genotype     directory                           sample
#1      Col salmon_quants   Col_RNAseq_meiocyte_Rep1_quant
#2      Col salmon_quants   Col_RNAseq_meiocyte_Rep2_quant
#3      Col salmon_quants   Col_RNAseq_meiocyte_Rep3_quant
#4    taf4b salmon_quants taf4b_RNAseq_meiocyte_Rep1_quant
#5    taf4b salmon_quants taf4b_RNAseq_meiocyte_Rep2_quant
#6    taf4b salmon_quants taf4b_RNAseq_meiocyte_Rep3_quant
#7     Bur0 salmon_quants  Bur0_RNAseq_meiocyte_Rep1_quant
#8     Bur0 salmon_quants  Bur0_RNAseq_meiocyte_Rep2_quant
#9     Bur0 salmon_quants  Bur0_RNAseq_meiocyte_Rep3_quant

# Specify paths to count files
files <- file.path(inDir, samples$genotype, samples$directory, samples$sample, "quant.sf")
# Set 1:#_samples
names(files) <- paste0("sample", 1:9)
all(file.exists(files))
#[1] TRUE

# Create a dataframe of transcript IDs and corresponding gene IDs
transID <- read.table(files[1], colClasses = c(NA, rep("NULL", 4)), header = T)
tx2gene <- data.frame(cbind(as.vector(transID[,1]), substr(transID[,1], 1, 9)))
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Import transcript-level counts, summarised at gene level
# (reads that map to transcript IDs with a common parent gene ID are pooled)
library(readr)
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
print(names(txi))
#[1] "abundance"           "counts"              "length"             
#[4] "countsFromAbundance"

# Import transcript-level counts, summarised at transcript level with "txOut = TRUE"
txi.tx <- tximport(files, type = "salmon", txOut = TRUE, tx2gene = tx2gene)
# Then summarise to gene level
txi.sum <- summarizeToGene(txi.tx, tx2gene)
# These two approaches should produce identical results
print(all.equal(txi$counts, txi.sum$counts))
#[1] TRUE

print(head(txi$counts))
#          sample1  sample2 sample3 sample4 sample5 sample6 sample7  sample8
#AT1G01010       2  2.00000       5       0       0       0       6  8.00000
#AT1G01020      52 29.12447     171       8       3       6     100 24.00000
#AT1G01030       1  0.00000       1       0       0       0       0  0.00000
#AT1G01040       2  9.00000      12      21       7      18      21 14.00000
#AT1G01050      11 25.00000      22      30       5       7     262 47.00000
#AT1G01060      10  7.00000       7       9       8       2      81 25.00434
#          sample9
#AT1G01010       5
#AT1G01020      59
#AT1G01030       0
#AT1G01040      16
#AT1G01050      87
#AT1G01060      41


## Use with downstream Bioconductor differential expression package
## DESeq2: http://www.bioconductor.org/help/workflows/rnaseqGene/

library(DESeq2)
print(packageVersion("DESeq2"))
#[1] ‘1.16.1’
sampleTable <- data.frame(sample = c("Col meiocyte RNA-seq Rep1", "Col meiocyte RNA-seq Rep2", "Col meiocyte RNA-seq Rep3",
                                     "taf4b meiocyte RNA-seq Rep1", "taf4b meiocyte RNA-seq Rep2", "taf4b meiocyte RNA-seq Rep3",
                                     "Bur0 meiocyte RNA-seq Rep1", "Bur0 meiocyte RNA-seq Rep2", "Bur0 meiocyte RNA-seq Rep3"),
                          condition = factor(rep.int(c("Col", "taf4b", "Bur0"), times = c(3, 3, 3))))
rownames(sampleTable) <- colnames(txi$counts)
print(sampleTable)
#                             sample condition
#sample1   Col meiocyte RNA-seq Rep1       Col
#sample2   Col meiocyte RNA-seq Rep2       Col
#sample3   Col meiocyte RNA-seq Rep3       Col
#sample4 taf4b meiocyte RNA-seq Rep1     taf4b
#sample5 taf4b meiocyte RNA-seq Rep2     taf4b
#sample6 taf4b meiocyte RNA-seq Rep3     taf4b
#sample7  Bur0 meiocyte RNA-seq Rep1      Bur0
#sample8  Bur0 meiocyte RNA-seq Rep2      Bur0
#sample9  Bur0 meiocyte RNA-seq Rep3      Bur0

dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

# Pre-filter the dataset
print(nrow(dds))
#[1] 27586
# Retain only rows that have more than a single count across all samples
dds <- dds[rowSums(counts(dds)) > 1,]
print(nrow(dds))
#[1] 24117


## The rlog and variance stabilizing transformations
# see http://www.bioconductor.org/help/workflows/rnaseqGene/#the-rlog-and-variance-stabilizing-transformations

rld <- rlog(dds, blind = FALSE)
print(head(assay(rld), 3))

vsd <- vst(dds, blind = FALSE)
print(head(assay(vsd), 3))

# Visualise the effect of transformation
library(dplyr)
library(ggplot2)
library(hexbin)

# For the log2 approach, estimate size factors to account for sequencing depth
# Sequencing-depth correction is done automatically for rlog and vst
dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized = T)[,1:2]+1)) %>%
    mutate(transformation = "log2(normalized counts + 1)"),
  as_data_frame(assay(rld)[,1:2]) %>% mutate(transformation = "rlog"),
  as_data_frame(assay(vsd)[,1:2]) %>% mutate(transformation = "vst"))

colnames(df)[1:2] <- c("Sample 1 transformed counts", "Sample 2 transformed counts")

plot_transformed_counts <- ggplot(df, aes(x = `Sample 1 transformed counts`, y = `Sample 2 transformed counts`)) +
                             geom_hex(bins = 80) + coord_fixed() + facet_grid(. ~ transformation) +
                             labs(fill = "Occurrences") +
                             theme_classic()
                             #theme(panel.border = element_rect(colour = "black", fill = NA, size = 1))
ggsave(plot_transformed_counts,
       file = paste0(plotDir, "Sample1_vs_Sample2_transformed_counts_log2countsPlus1_rlog_vst.pdf"))


## Sample distances

# Sample distances using the rlog-transformed counts
sampleDists <- dist(t(assay(rld)))
print(sampleDists)

library(pheatmap)
library(RColorBrewer)

# Heatmap of sample distances using the rlog-transformed counts
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- c("Col meiocyte RNA-seq Rep1", "Col meiocyte RNA-seq Rep2", "Col meiocyte RNA-seq Rep3",
                                "taf4b meiocyte RNA-seq Rep1", "taf4b meiocyte RNA-seq Rep2", "taf4b meiocyte RNA-seq Rep3",
                                "Bur0 meiocyte RNA-seq Rep1", "Bur0 meiocyte RNA-seq Rep2", "Bur0 meiocyte RNA-seq Rep3")
colnames(sampleDistMatrix) <- NULL
mycols <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pdf(paste0(plotDir, "sample_distances_heatmap_rlog.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = mycols)
dev.off()

# Sample distances using the Poisson Distance
library(PoiClaClu)
poisd <- PoissonDistance(t(counts(dds)))
print(poisd)
#Value of alpha used to transform data:  0.515102
#This type of normalization was performed: mle
#Dissimilarity computed for  9  observations.

# Heatmap
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- c("Col meiocyte RNA-seq Rep1", "Col meiocyte RNA-seq Rep2", "Col meiocyte RNA-seq Rep3",
                                    "taf4b meiocyte RNA-seq Rep1", "taf4b meiocyte RNA-seq Rep2", "taf4b meiocyte RNA-seq Rep3",
                                    "Bur0 meiocyte RNA-seq Rep1", "Bur0 meiocyte RNA-seq Rep2", "Bur0 meiocyte RNA-seq Rep3")
colnames(samplePoisDistMatrix) <- NULL
pdf(paste0(plotDir, "sample_distances_heatmap_Poisson.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = mycols)
dev.off()

# PCA plots for visualising sample-to-sample distances
PCAplot_rlog <- DESeq2::plotPCA(rld, intgroup = c("condition", "sample")) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(PCAplot_rlog,
       file = paste0(plotDir, "PCAplot_rlog.pdf"), width = 20, height = 20, units = "cm")

PCAplot_rlog_data <- DESeq2::plotPCA(rld, intgroup = c("condition", "sample"), returnData = T)
PCAplot_rlog_data
# Obtain percentage variance explained by PC1 and PC2 for plotting using ggplot2
percentVar <- round(100 * attr(PCAplot_rlog_data, "percentVar"))

PCAggplot_rlog <- ggplot(PCAplot_rlog_data, aes(x = PC1, y = PC2,
                                                shape = condition,
                                                colour = sample)) +
                    geom_point(size = 3) +
                    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
                    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
                    theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                    theme_classic() +
                    theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                    coord_fixed()
ggsave(PCAggplot_rlog,
       file = paste0(plotDir, "PCAggplot_rlog.pdf"), height = 20, width = 20, units = "cm")

# MDS (multi-dimensional scaling) plots for visualising sample-to-sample distances
# rlog-transformed counts
mds <- as.data.frame(colData(rld)) %>%
         cbind(cmdscale(sampleDistMatrix))
MDSplot_rlog <- ggplot(mds, aes(x = `1`, y = `2`,
                                shape = condition,
                                colour = sample)) +
                  geom_point(size = 3) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(MDSplot_rlog,
       file = paste0(plotDir, "MDSplot_rlog.pdf"), height = 20, width = 20, units = "cm")

# Poisson distance
mdsPois <- as.data.frame(colData(dds)) %>%
             cbind(cmdscale(samplePoisDistMatrix))
MDSplot_Pois <- ggplot(mdsPois, aes(x = `1`, y = `2`,
                                shape = condition,
                                colour = sample)) +
                  geom_point(size = 3) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(MDSplot_Pois,
       file = paste0(plotDir, "MDSplot_Pois.pdf"), height = 20, width = 20, units = "cm")

# Gene clustering
# Clustering of 20 genes with greatest rlog-transformed variance across samples
library(genefilter)
topVarGenes <- head(order(rowVars(assay(rld)), decreasing = T), 20)
mat <- assay(rld)[topVarGenes, ]
mat <- mat - rowMeans(mat)
anno <- as.data.frame(colData(rld)[, c("sample", "condition")])
pdf(paste0(plotDir, "gene_clustering_rld_topVar20.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(mat, annotation_col = anno)
dev.off()



## Differential expression analysis

# The DESeq developers recommend that the DESeq() function is run to fit a model using samples from all groups,
# followed by use of the "contrast" argument of the results() function to make comparisons between two groups.
# "The model fit by DESeq estimates a single dispersion parameter for each gene, which defines how far we expect
# the observed count for a sample will be from the mean value from the model given its size factor and its condition group."
# (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
# However, given that Col replicates have high within-group variability, as revealed by the above sample-distance analyses,
# greater sensitivity would be achieved by subsetting the DESeqDataSet to only samples from two groups (Col and another)
# before fitting a model by running the DESeq() function.
# This would prevent inflation of the per-gene dispersion estimate.


## taf4b vs Col

# Re-import DESeqDataSet to redefine "dds"
dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

# Subset DESeqDataSet to only samples from the two conditions to be contrasted
dds_taf4bVCol <- dds[ , dds$condition == "taf4b" | dds$condition == "Col"]
# Remove empty levels
dds_taf4bVCol$condition <- droplevels(dds_taf4bVCol$condition)
dds_taf4bVCol$sample <- droplevels(dds_taf4bVCol$sample)

# Pre-filter the datasets
print(nrow(dds_taf4bVCol))
#[1] 27586
# Retain only rows that have more than a single count across all samples
dds_taf4bVCol <- dds_taf4bVCol[rowSums(counts(dds_taf4bVCol)) > 1,]
print(nrow(dds_taf4bVCol))
#[1] 22179

# Contrast: taf4b vs Col 
dds_taf4bVCol <- DESeq(dds_taf4bVCol)
res_taf4bVCol_.1 <- results(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), alpha = 0.1)
print(mcols(res_taf4bVCol_.1, use.names = T))
#DataFrame with 6 rows and 2 columns
#                       type                                    description
#                <character>                                    <character>
#baseMean       intermediate      mean of normalized counts for all samples
#log2FoldChange      results log2 fold change (MLE): condition taf4b vs Col
#lfcSE               results         standard error: condition taf4b vs Col
#stat                results         Wald statistic: condition taf4b vs Col
#pvalue              results      Wald test p-value: condition taf4b vs Col
#padj                results                           BH adjusted p-values

print(summary(res_taf4bVCol_.1))
#out of 22179 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 528, 2.4% 
#LFC < 0 (down)   : 1205, 5.4% 
#outliers [1]     : 19, 0.086% 
#low counts [2]   : 8170, 37% 
#(mean count < 3)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_taf4bVCol_.1$padj < 0.1))
#FALSE  TRUE 
#12257  1733
print(sum(!is.na(res_taf4bVCol_.1$padj)))
#[1] 13990

res_taf4bVCol_.05 <- results(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), alpha = 0.05)
print(summary(res_taf4bVCol_.05))
#out of 22179 with nonzero total read count
#adjusted p-value < 0.05
#LFC > 0 (up)     : 410, 1.8% 
#LFC < 0 (down)   : 974, 4.4% 
#outliers [1]     : 19, 0.086% 
#low counts [2]   : 8600, 39% 
#(mean count < 4)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_taf4bVCol_.05$padj < 0.05))
#FALSE  TRUE 
#12176  1384

res_taf4bVCol_.01 <- results(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), alpha = 0.01)
print(summary(res_taf4bVCol_.01))
#out of 22179 with nonzero total read count
#adjusted p-value < 0.01
#LFC > 0 (up)     : 256, 1.2% 
#LFC < 0 (down)   : 613, 2.8% 
#outliers [1]     : 19, 0.086% 
#low counts [2]   : 9890, 45% 
#(mean count < 6)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_taf4bVCol_.01$padj < 0.01))
#FALSE  TRUE 
#11401   869

res_taf4bVCol_LFC1 <- results(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), lfcThreshold = 1)
print(summary(res_taf4bVCol_LFC1))
#out of 22179 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 14, 0.063% 
#LFC < 0 (down)   : 218, 0.98% 
#outliers [1]     : 19, 0.086% 
#low counts [2]   : 9890, 45% 
#(mean count < 6)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_taf4bVCol_LFC1$padj < 0.1))
#FALSE  TRUE 
#12038   232

res_taf4bVCol_LFC2 <- results(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), lfcThreshold = 2)
print(summary(res_taf4bVCol_LFC2))
#out of 22179 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 2, 0.009% 
#LFC < 0 (down)   : 21, 0.095% 
#outliers [1]     : 19, 0.086% 
#low counts [2]   : 11179, 50% 
#(mean count < 10)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_taf4bVCol_LFC2$padj < 0.1))
#FALSE  TRUE 
#10958    23

# MA-plots
# These show the log2(fold changes) in gene expression values for a variable (e.g., treated vs untreated)
# over a the mean of normalized counts for all the samples in the DESeqDataSet

# DESeq2 provides for use of a "Bayesian procedure to moderate (or "shrink") log2 fold changes from genes
# with very low counts and highly variable counts"
# Before generating an MA-plot, the lfcShrink() function should be used to moderate the log2(fold changes) for the contrast
res_taf4bVCol_.1_lfcShrink <- lfcShrink(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), res = res_taf4bVCol_.1)
pdf(paste0(plotDir, "MAplot_res_taf4bVCol_.1_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_taf4bVCol_.1_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_taf4bVCol_.1_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

# The following MA-plot is generated without moderating the noisy log2(fold changes)
pdf(paste0(plotDir, "MAplot_res_taf4bVCol_.1.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_taf4bVCol_.1, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_taf4bVCol_.1, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

res_taf4bVCol_.05_lfcShrink <- lfcShrink(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), res = res_taf4bVCol_.05)
pdf(paste0(plotDir, "MAplot_res_taf4bVCol_.05_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_taf4bVCol_.05_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_taf4bVCol_.05_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

res_taf4bVCol_.01_lfcShrink <- lfcShrink(dds_taf4bVCol, contrast = c("condition", "taf4b", "Col"), res = res_taf4bVCol_.01)
pdf(paste0(plotDir, "MAplot_res_taf4bVCol_.01_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_taf4bVCol_.01_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_taf4bVCol_.01_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()


# Histograms of unadjusted and FDR-adjusted p-values
pdf(paste0(plotDir, "hist_res_taf4bVCol_.1_pvals_unadjusted.pdf"))
hist(res_taf4bVCol_.1$pvalue[res_taf4bVCol_.1$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()

pdf(paste0(plotDir, "hist_res_taf4bVCol_.1_pvals_BH_FDR_adjusted.pdf"))
hist(res_taf4bVCol_.1$padj[res_taf4bVCol_.1$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()


# Independent filtering
# Create bins
qs <- c(0, quantile(res_taf4bVCol_.1$baseMean[res_taf4bVCol_.1$baseMean > 0], 0:6/6))
# Bin genes by base mean (mean normalized count) using cut
bins <- cut(res_taf4bVCol_.1$baseMean, qs)
# Rename the levels of the bins using the middle point
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(res_taf4bVCol_.1$pvalue, bins, function(p)
                 mean(p < 0.05, na.rm = TRUE))
pdf(paste0(plotDir, "barplot_res_taf4bVCol_.1_fraction_small_pval_genes.pdf"))
barplot(fractionSig, xlab = "Mean normalized count",
                     ylab = "Fraction of small P values")
dev.off()


## Exporting results

# FDR < 0.1
res_taf4bVCol_.1_lfcShrink_Chr <- res_taf4bVCol_.1_lfcShrink[!grepl("ATM", rownames(res_taf4bVCol_.1_lfcShrink)) &
                                                             !grepl("ATC", rownames(res_taf4bVCol_.1_lfcShrink)), ]

# Subset results table to significant (FDR < 0.1) down-regulated genes on chromosomes in taf4b
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downReg <- subset(res_taf4bVCol_.1_lfcShrink_Chr, padj < 0.1 & log2FoldChange < 0)
print(dim(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downReg))
#[1] 1203    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSorted <- res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downReg[order(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.1) up-regulated genes on chromosomes in taf4b
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upReg <- subset(res_taf4bVCol_.1_lfcShrink_Chr, padj < 0.1 & log2FoldChange > 0)
print(dim(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upReg))
#[1] 355   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSorted <- res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upReg[order(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF <- as.data.frame(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSorted)
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF <- as.data.frame(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSorted)
write.table(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF.txt"))
write.table(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF.txt"))

res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs <- rownames(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF)
res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs <- rownames(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF)
write.table(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs.txt"))
write.table(res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs.txt"))


# FDR < 0.05
res_taf4bVCol_.05_lfcShrink_Chr <- res_taf4bVCol_.05_lfcShrink[!grepl("ATM", rownames(res_taf4bVCol_.05_lfcShrink)) &
                                                               !grepl("ATC", rownames(res_taf4bVCol_.05_lfcShrink)), ]

# Subset results table to significant (FDR < 0.05) down-regulated genes on chromosomes in taf4b
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downReg <- subset(res_taf4bVCol_.05_lfcShrink_Chr, padj < 0.05 & log2FoldChange < 0)
print(dim(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downReg))
#[1] 972    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSorted <- res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downReg[order(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.05) up-regulated genes on chromosomes in taf4b
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upReg <- subset(res_taf4bVCol_.05_lfcShrink_Chr, padj < 0.05 & log2FoldChange > 0)
print(dim(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upReg))
#[1] 246   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSorted <- res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upReg[order(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF <- as.data.frame(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSorted)
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF <- as.data.frame(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSorted)
write.table(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF.txt"))
write.table(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF.txt"))

res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs <- rownames(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF)
res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs <- rownames(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF)
write.table(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs.txt"))
write.table(res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs.txt"))
                                                                                           

# FDR < 0.01
res_taf4bVCol_.01_lfcShrink_Chr <- res_taf4bVCol_.01_lfcShrink[!grepl("ATM", rownames(res_taf4bVCol_.01_lfcShrink)) &
                                                               !grepl("ATC", rownames(res_taf4bVCol_.01_lfcShrink)), ]

# Subset results table to significant (FDR < 0.01) down-regulated genes on chromosomes in taf4b
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downReg <- subset(res_taf4bVCol_.01_lfcShrink_Chr, padj < 0.01 & log2FoldChange < 0)
print(dim(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downReg))
#[1] 612    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSorted <- res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downReg[order(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.01) up-regulated genes on chromosomes in taf4b
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upReg <- subset(res_taf4bVCol_.01_lfcShrink_Chr, padj < 0.01 & log2FoldChange > 0)
print(dim(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upReg))
#[1] 131   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSorted <- res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upReg[order(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF <- as.data.frame(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSorted)
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF <- as.data.frame(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSorted)
write.table(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF.txt"))
write.table(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF,
            file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF.txt"))

res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs <- rownames(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF)
res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs <- rownames(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF)
write.table(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs.txt"))
write.table(res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs,
            file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs.txt"))



## Bur0 vs Col

# Re-import DESeqDataSet to redefine "dds"
dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

# Subset DESeqDataSet to only samples from the two conditions to be contrasted
dds_Bur0VCol <- dds[ , dds$condition == "Bur0" | dds$condition == "Col"]
# Remove empty levels
dds_Bur0VCol$condition <- droplevels(dds_Bur0VCol$condition)
dds_Bur0VCol$sample <- droplevels(dds_Bur0VCol$sample)

# Pre-filter the datasets
print(nrow(dds_Bur0VCol))
#[1] 27586
# Retain only rows that have more than a single count across all samples
dds_Bur0VCol <- dds_Bur0VCol[rowSums(counts(dds_Bur0VCol)) > 1,]
print(nrow(dds_Bur0VCol))
#[1] 22703

# Contrast: Bur0 vs Col 
dds_Bur0VCol <- DESeq(dds_Bur0VCol)
res_Bur0VCol_.1 <- results(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), alpha = 0.1)
print(mcols(res_Bur0VCol_.1, use.names = T))
#DataFrame with 6 rows and 2 columns
#                       type                                   description
#                <character>                                   <character>
#baseMean       intermediate     mean of normalized counts for all samples
#log2FoldChange      results log2 fold change (MLE): condition Bur0 vs Col
#lfcSE               results         standard error: condition Bur0 vs Col
#stat                results         Wald statistic: condition Bur0 vs Col
#pvalue              results      Wald test p-value: condition Bur0 vs Col
#padj                results                          BH adjusted p-values

print(summary(res_Bur0VCol_.1))
#out of 22703 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 667, 2.9% 
#LFC < 0 (down)   : 2125, 9.4% 
#outliers [1]     : 16, 0.07% 
#low counts [2]   : 6603, 29% 
#(mean count < 3)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_Bur0VCol_.1$padj < 0.1))
#FALSE  TRUE 
#13292  2792
print(sum(!is.na(res_Bur0VCol_.1$padj)))
#[1] 16084

res_Bur0VCol_.05 <- results(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), alpha = 0.05)
print(summary(res_Bur0VCol_.05))
#out of 22703 with nonzero total read count
#adjusted p-value < 0.05
#LFC > 0 (up)     : 392, 1.7% 
#LFC < 0 (down)   : 1719, 7.6% 
#outliers [1]     : 16, 0.07% 
#low counts [2]   : 7043, 31% 
#(mean count < 3)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_Bur0VCol_.05$padj < 0.05))
#FALSE  TRUE 
#13533  2111

res_Bur0VCol_.01 <- results(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), alpha = 0.01)
print(summary(res_Bur0VCol_.01))
#out of 22703 with nonzero total read count
#adjusted p-value < 0.01
#LFC > 0 (up)     : 154, 0.68% 
#LFC < 0 (down)   : 1123, 4.9% 
#outliers [1]     : 16, 0.07% 
#low counts [2]   : 7923, 35% 
#(mean count < 5)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_Bur0VCol_.01$padj < 0.01))
#FALSE  TRUE 
#13487  1277

res_Bur0VCol_LFC1 <- results(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), lfcThreshold = 1)
print(summary(res_Bur0VCol_LFC1))
#out of 22703 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 69, 0.3% 
#LFC < 0 (down)   : 537, 2.4% 
#outliers [1]     : 16, 0.07% 
#low counts [2]   : 7483, 33% 
#(mean count < 4)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_Bur0VCol_LFC1$padj < 0.1))
#FALSE  TRUE 
#14598   606

res_Bur0VCol_LFC2 <- results(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), lfcThreshold = 2)
print(summary(res_Bur0VCol_LFC2))
#out of 22703 with nonzero total read count
#adjusted p-value < 0.1
#LFC > 0 (up)     : 26, 0.11% 
#LFC < 0 (down)   : 122, 0.54% 
#outliers [1]     : 16, 0.07% 
#low counts [2]   : 8363, 37% 
#(mean count < 5)
#[1] see 'cooksCutoff' argument of ?results
#[2] see 'independentFiltering' argument of ?results
print(table(res_Bur0VCol_LFC2$padj < 0.1))
#FALSE  TRUE 
#14176   148

# MA-plots
# These show the log2(fold changes) in gene expression values for a variable (e.g., treated vs untreated)
# over a the mean of normalized counts for all the samples in the DESeqDataSet

# DESeq2 provides for use of a "Bayesian procedure to moderate (or "shrink") log2 fold changes from genes
# with very low counts and highly variable counts"
# Before generating an MA-plot, the lfcShrink() function should be used to moderate the log2(fold changes) for the contrast
res_Bur0VCol_.1_lfcShrink <- lfcShrink(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), res = res_Bur0VCol_.1)
pdf(paste0(plotDir, "MAplot_res_Bur0VCol_.1_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_Bur0VCol_.1_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_Bur0VCol_.1_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

# The following MA-plot is generated without moderating the noisy log2(fold changes)
pdf(paste0(plotDir, "MAplot_res_Bur0VCol_.1.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_Bur0VCol_.1, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_Bur0VCol_.1, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

res_Bur0VCol_.05_lfcShrink <- lfcShrink(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), res = res_Bur0VCol_.05)
pdf(paste0(plotDir, "MAplot_res_Bur0VCol_.05_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_Bur0VCol_.05_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_Bur0VCol_.05_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

res_Bur0VCol_.01_lfcShrink <- lfcShrink(dds_Bur0VCol, contrast = c("condition", "Bur0", "Col"), res = res_Bur0VCol_.01)
pdf(paste0(plotDir, "MAplot_res_Bur0VCol_.01_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_Bur0VCol_.01_lfcShrink, ylim = c(-10.5, 10.5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_Bur0VCol_.01_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()


# Histograms of unadjusted and FDR-adjusted p-values
pdf(paste0(plotDir, "hist_res_Bur0VCol_.1_pvals_unadjusted.pdf"))
hist(res_Bur0VCol_.1$pvalue[res_Bur0VCol_.1$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()

pdf(paste0(plotDir, "hist_res_Bur0VCol_.1_pvals_BH_FDR_adjusted.pdf"))
hist(res_Bur0VCol_.1$padj[res_Bur0VCol_.1$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()


# Independent filtering
# Create bins
qs <- c(0, quantile(res_Bur0VCol_.1$baseMean[res_Bur0VCol_.1$baseMean > 0], 0:6/6))
# Bin genes by base mean (mean normalized count) using cut
bins <- cut(res_Bur0VCol_.1$baseMean, qs)
# Rename the levels of the bins using the middle point
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(res_Bur0VCol_.1$pvalue, bins, function(p)
                 mean(p < 0.05, na.rm = TRUE))
pdf(paste0(plotDir, "barplot_res_Bur0VCol_.1_fraction_small_pval_genes.pdf"))
barplot(fractionSig, xlab = "Mean normalized count",
                     ylab = "Fraction of small P values")
dev.off()


## Exporting results

# FDR < 0.1
res_Bur0VCol_.1_lfcShrink_Chr <- res_Bur0VCol_.1_lfcShrink[!grepl("ATM", rownames(res_Bur0VCol_.1_lfcShrink)) &
                                                           !grepl("ATC", rownames(res_Bur0VCol_.1_lfcShrink)), ]

# Subset results table to significant (FDR < 0.1) down-regulated genes on chromosomes in Bur0
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg <- subset(res_Bur0VCol_.1_lfcShrink_Chr, padj < 0.1 & log2FoldChange < 0)
print(dim(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg))
#[1] 2114    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSorted <- res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg[order(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.1) up-regulated genes on chromosomes in Bur0
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg <- subset(res_Bur0VCol_.1_lfcShrink_Chr, padj < 0.1 & log2FoldChange > 0)
print(dim(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg))
#[1] 652   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSorted <- res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg[order(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF <- as.data.frame(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSorted)
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF <- as.data.frame(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSorted)
write.table(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF.txt"))
write.table(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF.txt"))

res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs <- rownames(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegSortedDF)
res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs <- rownames(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegSortedDF)
write.table(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs.txt"))
write.table(res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs.txt"))


# FDR < 0.05
res_Bur0VCol_.05_lfcShrink_Chr <- res_Bur0VCol_.05_lfcShrink[!grepl("ATM", rownames(res_Bur0VCol_.05_lfcShrink)) &
                                                             !grepl("ATC", rownames(res_Bur0VCol_.05_lfcShrink)), ]

# Subset results table to significant (FDR < 0.05) down-regulated genes on chromosomes in Bur0
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg <- subset(res_Bur0VCol_.05_lfcShrink_Chr, padj < 0.05 & log2FoldChange < 0)
print(dim(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg))
#[1] 1711    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSorted <- res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg[order(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.05) up-regulated genes on chromosomes in Bur0
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg <- subset(res_Bur0VCol_.05_lfcShrink_Chr, padj < 0.05 & log2FoldChange > 0)
print(dim(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg))
#[1] 379   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSorted <- res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg[order(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF <- as.data.frame(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSorted)
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF <- as.data.frame(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSorted)
write.table(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF.txt"))
write.table(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF.txt"))

res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs <- rownames(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegSortedDF)
res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs <- rownames(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegSortedDF)
write.table(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs.txt"))
write.table(res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs.txt"))


# FDR < 0.01
res_Bur0VCol_.01_lfcShrink_Chr <- res_Bur0VCol_.01_lfcShrink[!grepl("ATM", rownames(res_Bur0VCol_.01_lfcShrink)) &
                                                             !grepl("ATC", rownames(res_Bur0VCol_.01_lfcShrink)), ]

# Subset results table to significant (FDR < 0.01) down-regulated genes on chromosomes in Bur0
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg <- subset(res_Bur0VCol_.01_lfcShrink_Chr, padj < 0.01 & log2FoldChange < 0)
print(dim(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg))
#[1] 1119    5
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSorted <- res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg[order(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg$log2FoldChange),]
# Subset results table to significant (FDR < 0.01) up-regulated genes on chromosomes in Bur0
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg <- subset(res_Bur0VCol_.01_lfcShrink_Chr, padj < 0.01 & log2FoldChange > 0)
print(dim(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg))
#[1] 145   5
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSorted <- res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg[order(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF <- as.data.frame(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSorted)
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF <- as.data.frame(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSorted)
write.table(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF.txt"))
write.table(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF,
            file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF.txt"))

res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs <- rownames(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegSortedDF)
res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs <- rownames(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegSortedDF)
write.table(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs.txt"))
write.table(res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs,
            file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs.txt"))


## Generate plots of log-transformed normalized counts at differentially expressed genes in taf4b 

taf4b_downReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/taf4b/DEGdownReg_counts/"
taf4b_upReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/taf4b/DEGupReg_counts/"
Bur0_downReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/Bur0/DEGdownReg_counts/"
Bur0_upReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/Bur0/DEGupReg_counts/"
taf4b_Bur0_downReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/taf4b_Bur0/DEGdownReg_counts/"
taf4b_Bur0_upReg_plotDir <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Feng/DESeq2_analysis_03/plots/taf4b_Bur0/DEGupReg_counts/"

library(ggplot2)
library(ggbeeswarm)
library(parallel)
library(org.At.tair.db)
TAIRsymbol <- org.At.tairSYMBOL
# Get TAIR gene identifiers that are mapped to a gene symbol
mapped_geneIDs <- mappedkeys(TAIRsymbol)

dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs.txt"))$x)
taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs.txt"))$x)
Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs.txt"))$x)
Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs.txt"))$x)

taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs.txt"))$x)
taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs.txt"))$x)
Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs.txt"))$x)
Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs.txt"))$x)

taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs.txt"))$x)
taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs.txt"))$x)
Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs.txt"))$x)
Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs.txt"))$x)

# Count-plotting function
countPlotFun <- function(geneID, contrast, plotDir) {
  geneCounts <- plotCounts(dds, gene = geneID, intgroup = c("condition", "sample"),
                         normalized = T, transform = T, returnData = T)
  geneCounts$condition <- factor(geneCounts$condition, levels = c("Col", "taf4b", "Bur0"))
  geneCountsPlot <- ggplot(geneCounts, aes(x = condition, y = count, colour = sample)) +
                      scale_y_log10(breaks = pretty(geneCounts$count)) +
                      geom_beeswarm(cex = 3) +
                      xlab("Genotype") +
                      ylab("log-transformed normalized count") +
                      labs(colour = "Sample") +
                      theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                      theme_classic() +
                      theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                      ggtitle(paste0(geneID, " (", TAIRsymbol[[geneID]][1], ")")) +
                      theme(plot.title = element_text(hjust = 0.5))
  ggsave(geneCountsPlot,
         file = paste0(plotDir, contrast, "_", geneID, "_", gsub("/", "|", TAIRsymbol[[geneID]][1]), "_normalized_counts.pdf"),
         width = 16, height = 10, units = "cm")
}

# FDR < 0.01
mclapply(seq_along(taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs[x],
               contrast = "taf4bVCol_.01_lfcShrink_Chr_Sig.01_downReg",
               plotDir = taf4b_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs[x],
               contrast = "taf4bVCol_.01_lfcShrink_Chr_Sig.01_upReg",
               plotDir = taf4b_upReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs[x],
               contrast = "Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg",
               plotDir = Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs[x],
               contrast = "Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg",
               plotDir = Bur0_upReg_plotDir)
}, mc.cores = 24)

# FDR < 0.05
mclapply(seq_along(taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs[x],
               contrast = "taf4bVCol_.05_lfcShrink_Chr_Sig.05_downReg",
               plotDir = taf4b_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs[x],
               contrast = "taf4bVCol_.05_lfcShrink_Chr_Sig.05_upReg",
               plotDir = taf4b_upReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs[x],
               contrast = "Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg",
               plotDir = Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs[x],
               contrast = "Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg",
               plotDir = Bur0_upReg_plotDir)
}, mc.cores = 24)

# FDR < 0.1
mclapply(seq_along(taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs[x],
               contrast = "taf4bVCol_.1_lfcShrink_Chr_Sig.1_downReg",
               plotDir = taf4b_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs[x],
               contrast = "taf4bVCol_.1_lfcShrink_Chr_Sig.1_upReg",
               plotDir = taf4b_upReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs[x],
               contrast = "Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg",
               plotDir = Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs), function(x) {
  countPlotFun(geneID = Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs[x],
               contrast = "Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg",
               plotDir = Bur0_upReg_plotDir)
}, mc.cores = 24)


# Identify genes that are down-regulated or up-regulated in both taf4b and Bur0

taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs_Bool <- taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs %in% Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs
taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs <- taf4bVCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs[taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs_Bool == TRUE]
taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs_Bool <- taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs %in% Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs
taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs <- taf4bVCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs[taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs_Bool == TRUE]

taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs_Bool <- taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs %in% Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs
taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs <- taf4bVCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs[taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs_Bool == TRUE]
taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs_Bool <- taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs %in% Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs
taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs <- taf4bVCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs[taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs_Bool == TRUE]

taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs_Bool <- taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs %in% Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs
taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs <- taf4bVCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs[taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs_Bool == TRUE]
taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs_Bool <- taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs %in% Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs
taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs <- taf4bVCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs[taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs_Bool == TRUE]

write.table(taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs.txt"))
write.table(taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs.txt"))
write.table(taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs.txt"))
write.table(taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs.txt"))
write.table(taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs.txt"))
write.table(taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs,
            file = paste0(outDir, "taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs.txt"))

# FDR < 0.01
mclapply(seq_along(taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_downReg",
               plotDir = taf4b_Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.01_lfcShrink_Chr_Sig.01_upReg",
               plotDir = taf4b_Bur0_upReg_plotDir)
}, mc.cores = 24)

# FDR < 0.05
mclapply(seq_along(taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_downReg",
               plotDir = taf4b_Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.05_lfcShrink_Chr_Sig.05_upReg",
               plotDir = taf4b_Bur0_upReg_plotDir)
}, mc.cores = 24)

# FDR < 0.1
mclapply(seq_along(taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_downReg",
               plotDir = taf4b_Bur0_downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs), function(x) {
  countPlotFun(geneID = taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upRegGeneIDs[x],
               contrast = "taf4bVCol_Bur0VCol_.1_lfcShrink_Chr_Sig.1_upReg",
               plotDir = taf4b_Bur0_upReg_plotDir)
}, mc.cores = 24)



